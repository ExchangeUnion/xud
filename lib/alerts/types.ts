import { AlertType, ChannelSide } from '../constants/enums';

type BaseAlert = {
  type: AlertType;
  message: string;
  date: number;
};

export type BalanceAlertEvent = {
  /** The total balance of the channel when the alert is triggered. */
  totalBalance: number;
  /** The side of the balance either local or remote. */
  side: ChannelSide;
  /** The balance that triggered the alert. */
  sideBalance: number;
  /** The alert threshold in percentage, e.g. 10 means %10. */
  bound: number;
  /** The currency of the channel. */
  currency: string;
};
export type BalanceAlert = BaseAlert & BalanceAlertEvent;

export type Alert = BalanceAlert;
