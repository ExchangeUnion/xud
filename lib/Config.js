const fs = require('fs');
const os = require('os');
const toml = require('toml');
const util = require('./utils/utils');
const enums = require('./constants/enums');


class Config {
  constructor(args) {
    this.args = args || null;
    const platform = os.platform();
    let lndDatadir;
    switch (platform) {
      case 'win32': { // windows
        const homeDir = process.env.LOCALAPPDATA;
        this.xuDir = `${homeDir}/Xud/`;
        lndDatadir = `${homeDir}/Lnd/`; // default lnd directory location
        break;
      }
      case 'darwin': { // mac
        const homeDir = process.env.HOME;
        this.xuDir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/Library/Application Support/Lnd/`;
        break;
      }
      default: { // linux
        const homeDir = process.env.HOME;
        this.xuDir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/.lnd/`;
        break;
      }
    }

    // default configuration
    this.p2p = {
      listen: true,
      port: 8885, // X = 88, U = 85 in ASCII
    };
    this.db = {
      host: 'localhost',
      port: 3306,
      username: 'xud',
      password: null,
      database: 'xud',
      dialect: 'mysql',
      pool: { max: 100 },
      operatorsAliases: false,
    };
    this.testDb = {
      ...this.db,
      database: 'xud_test',
    };
    this.raiden = {
      port: 5001,
    };
    this.rpcPort = 8886;
    this.lnd = {
      datadir: lndDatadir,
      rpcProtoPath: 'lndrpc.proto',
    };
    this.swapProtocols = {
      [enums.swapProtocols.LND]: {
        enable: true,
      },
      [enums.swapProtocols.RAIDEN]: {
        enable: true,
      },
    };
  }

  async load() {
    if (!fs.existsSync(this.xuDir)) {
      fs.mkdirSync(this.xuDir);
    } else if (fs.existsSync(`${this.xuDir}xud.conf`)) {
      const configText = fs.readFileSync(`${this.xuDir}xud.conf`);
      try {
        const props = toml.parse(configText);

        // copy parsed json properties from config file to this config object
        util.copyObject(props, this);
      } catch (e) {
        throw new Error(`Parsing error on line ${e.line}, column ${e.column
        }: ${e.message}`);
      }
    }

    if (this.args) {
      // override our config file with command line arguments
      util.copyObject(this.args, this);
      this.args = null;
    }
  }

  get(key) {
    return this.props[key];
  }

  isLndEnabled() {
    return this.swapProtocols[enums.swapProtocols.LND].enable;
  }

  isRaidenEnabled() {
    return this.swapProtocols[enums.swapProtocols.RAIDEN].enable;
  }
}

module.exports = Config;
