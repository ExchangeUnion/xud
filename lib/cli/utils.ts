import colors from 'colors/safe';
import { accessSync, watch } from 'fs';
import os from 'os';
import path from 'path';

const SATOSHIS_PER_COIN = 10 ** 8;

export function getDefaultCertPath() {
  switch (os.platform()) {
    case 'win32': {
      const homeDir = process.env.LOCALAPPDATA!;
      return path.join(homeDir, 'Xud', 'tls.cert');
    }
    case 'darwin': {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
    default: {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
  }
}

export function shorten(key: string, length = 10) {
  if (key.length <= length * 2 + 3) {
    return key;
  }
  return `${key.slice(0, length)}...${key.slice(key.length - length)}`;
}

export function trim(key: string, length = 10) {
  if (key.length <= length + 3) {
    return '';
  }
  return `${key.slice(0, length)}...`;
}

export const generateHeaders = (headers: string[]) => {
  return headers.map((header) => {
    return colors.blue(header);
  });
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number) => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number) => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Waits up to 5 seconds for the tls.cert file to be created in case this is the first time xud has been run. */
export const waitForCert = (certPath: string) => {
  return new Promise<void>((resolve, reject) => {
    try {
      accessSync(certPath);
      resolve();
    } catch (err) {
      if (err.code === 'ENOENT') {
        const certDir = path.dirname(certPath);
        const certFilename = path.basename(certPath);
        const fsWatcher = watch(certDir, (event, filename) => {
          if (event === 'change' && filename === certFilename) {
            clearTimeout(timeout);
            fsWatcher.close();
            resolve();
          }
        });
        const timeout = setTimeout(() => {
          fsWatcher.close();
          reject(`timed out waiting for cert to be created at ${certPath}`);
        }, 5000);
      } else {
        // we handle errors due to file not existing, otherwise reject
        reject(err);
      }
    }
  });
};

export function showSeed(seedMnemonicList: string[]) {
  const WORDS_PER_ROW = 4;
  const numberedMnemonic = seedMnemonicList.map((value, index) => {
    return `${index >= 9 ? '' : ' '}${index + 1}. ${value.padEnd(10)}`;
  });
  console.log('----------------------BEGIN XUD SEED---------------------');
  for (let n = 0; n < seedMnemonicList.length / WORDS_PER_ROW; n += 1) {
    console.log(numberedMnemonic.slice(n * WORDS_PER_ROW, (n + 1) * WORDS_PER_ROW).join(' '));
  }
  console.log('-----------------------END XUD SEED----------------------\n');

  console.log(`
Please write down your 24 word mnemonic. It will allow you to recover your xud \
node key and on-chain funds for the initialized wallets listed above should you \
forget your password or lose your device. Off-chain funds in channels can NOT \
be recovered with it and must be backed up and recovered separately. Keep it \
somewhere safe, it is your ONLY backup in case of data loss.
`);
}
