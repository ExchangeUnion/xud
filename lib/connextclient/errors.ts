import errorCodesPrefix from '../constants/errorCodesPrefix';

const codesPrefix = errorCodesPrefix.CONNEXT;
const errorCodes = {
  CONNEXT_IS_DISABLED: codesPrefix.concat('.1'),
  INSUFFICIENT_BALANCE: codesPrefix.concat('.2'),
  TIMEOUT: codesPrefix.concat('.3'),
  INVALID: codesPrefix.concat('.4'),
  SERVER_ERROR: codesPrefix.concat('.5'),
  UNEXPECTED: codesPrefix.concat('.6'),
  INVALID_TOKEN_PAYMENT_RESPONSE: codesPrefix.concat('.9'),
  CONNEXT_HAS_NO_ACTIVE_CHANNELS: codesPrefix.concat('.11'),
  CONNEXT_CLIENT_NOT_INITIALIZED: codesPrefix.concat('.12'),
  CURRENCY_NOT_FOUND_BY_TOKENADDRESS: codesPrefix.concat('.13'),
  CURRENCY_MISSING: codesPrefix.concat('.14'),
  EXPIRY_MISSING: codesPrefix.concat('.15'),
  MISSING_SEED: codesPrefix.concat('.16'),
  INSUFFICIENT_COLLATERAL: codesPrefix.concat('.17'),
  NOT_FOUND: codesPrefix.concat('.18'),
  WITHDRAW_ADDRESS_MISSING: codesPrefix.concat('.19'),
};

const errors = {
  CONNEXT_IS_DISABLED: {
    message: 'connext is disabled',
    code: errorCodes.CONNEXT_IS_DISABLED,
  },
  CONNEXT_HAS_NO_ACTIVE_CHANNELS: {
    message: 'connext has no active channels',
    code: errorCodes.CONNEXT_HAS_NO_ACTIVE_CHANNELS,
  },
  CONNEXT_CLIENT_NOT_INITIALIZED: {
    message: 'connext client is not initialized',
    code: errorCodes.CONNEXT_CLIENT_NOT_INITIALIZED,
  },
  INSUFFICIENT_BALANCE: {
    message: 'insufficient balance to perform request',
    code: errorCodes.INSUFFICIENT_BALANCE,
  },
  TIMEOUT: {
    message: 'request timed out',
    code: errorCodes.TIMEOUT,
  },
  SERVER_ERROR: (statusCode: number, statusMessage?: string) => ({
    message: `connext server error ${statusCode}: ${statusMessage ?? ''}`,
    code: errorCodes.SERVER_ERROR,
  }),
  UNEXPECTED: (statusCode: number, statusMessage?: string) => ({
    message: `unexpected status from connext request ${statusCode}: ${statusMessage ?? ''}`,
    code: errorCodes.UNEXPECTED,
  }),
  NOT_FOUND: {
    message: 'connext returned not found response',
    code: errorCodes.NOT_FOUND,
  },
  WITHDRAW_ADDRESS_MISSING: {
    message: 'destination account for the withdrawal is missing',
    code: errorCodes.WITHDRAW_ADDRESS_MISSING,
  },
  INVALID_TOKEN_PAYMENT_RESPONSE: {
    message: 'connext TokenPaymentResponse is invalid',
    code: errorCodes.INVALID_TOKEN_PAYMENT_RESPONSE,
  },
  CURRENCY_NOT_FOUND_BY_TOKENADDRESS: (tokenAddress: string) => ({
    message: `could not find currency for tokenAddress ${tokenAddress}`,
    code: errorCodes.CURRENCY_NOT_FOUND_BY_TOKENADDRESS,
  }),
  CURRENCY_MISSING: {
    message: 'currency is missing',
    code: errorCodes.CURRENCY_MISSING,
  },
  EXPIRY_MISSING: {
    message: 'expiry is missing',
    code: errorCodes.EXPIRY_MISSING,
  },
  MISSING_SEED: {
    message: 'seed is missing',
    code: errorCodes.MISSING_SEED,
  },
  INSUFFICIENT_COLLATERAL: {
    message: 'channel collateralization in progress, please try again in ~1 minute',
    code: errorCodes.INSUFFICIENT_COLLATERAL,
  },
};

export { errorCodes };
export default errors;
