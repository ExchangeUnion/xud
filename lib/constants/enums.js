const enums = exports;

enums.swapProtocols = {
  LND: 'LND',
  RAIDEN: 'RAIDEN',
};

enums.orderingDirections = {
  DESC: 'DESC',
  ASC: 'ASC',
};
