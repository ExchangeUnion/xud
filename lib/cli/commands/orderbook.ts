import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';
import { Arguments, Argv } from 'yargs';
import { OrderBookRequest, OrderBookResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';
import { satsToCoinsStr } from '../utils';

const COLUMNS = [19, 19, 19, 19];
const HEADER = [
  { content: colors.green('Buy'), colSpan: 2 },
  { content: colors.red('Sell'), colSpan: 2 },
];
const SECONDARY_HEADER = [colors.green('Quantity'), colors.green('Price'), colors.red('Price'), colors.red('Quantity')];

let precision = 0;

const createTable = () => {
  const table = new Table({ colWidths: COLUMNS }) as HorizontalTable;
  table.push(HEADER);
  table.push(SECONDARY_HEADER);
  return table;
};

const displayOrderbook = (pairId: string, orderbook: OrderBookResponse.Buckets.AsObject) => {
  const table = createTable();
  const { buyBucketsList, sellBucketsList } = orderbook;
  const rowCount = Math.max(buyBucketsList.length, sellBucketsList.length);

  for (let n = 0; n < rowCount; n += 1) {
    const row: string[] = [];
    row[0] = buyBucketsList[n] ? satsToCoinsStr(buyBucketsList[n].quantity) : '';
    row[1] = buyBucketsList[n]?.price.toString() ?? '';
    if (row[1] === '0') {
      // instead of displaying zero, display the < symbol and the smallest price bucket
      // given the specified precision. for example if precision is 2, a zero value
      // would be converted to "<0.01"
      row[1] = `<${1 / 10 ** precision}`;
    }
    row[2] = sellBucketsList[n]?.price.toString() ?? '';
    row[3] = sellBucketsList[n] ? satsToCoinsStr(sellBucketsList[n].quantity) : '';
    table.push(row);
  }

  console.log(colors.underline(colors.bold(`\nTrading pair: ${pairId}`)));
  console.log(table.toString());
};

export const displayTables = (orderbooks: OrderBookResponse.AsObject) => {
  orderbooks.bucketsMap.forEach((val) => {
    displayOrderbook(val[0], val[1]);
  });
};

export const command = 'orderbook [pair_id] [precision]';

export const describe = 'display the order book, with orders aggregated per price point';

export const builder = (argv: Argv) =>
  argv
    .option('pair_id', {
      describe: 'trading pair for which to retrieve the order book',
      type: 'string',
    })
    .option('precision', {
      describe: 'the number of digits following the decimal point',
      type: 'number',
      default: 5,
    })
    .option('limit', {
      describe: 'the number of digits following the decimal point',
      type: 'number',
      default: 8,
    })
    .example('$0 orderbook', 'display the order books for all trading pairs')
    .example('$0 orderbook LTC/BTC', 'display the LTC/BTC order book')
    .example('$0 orderbook LTC/BTC 2', 'display the LTC/BTC order book with 2 decimal precision')
    .example('$0 orderbook --precision 2', 'display the order books for all trading pairs with 2 decimal precision')
    .example(
      '$0 orderbook BTC/USDT -2 --limit 10',
      'display the USDT/BTC order book with up to 10 buckets using hundreds place precision',
    );

export const handler = async (argv: Arguments<any>) => {
  const request = new OrderBookRequest();
  const pairId = argv.pair_id ? argv.pair_id.toUpperCase() : undefined;
  request.setPairId(pairId);
  request.setPrecision(argv.precision);
  precision = argv.precision; // store the precision to a global variable in case we need it for display purposes later
  request.setLimit(argv.limit);
  (await loadXudClient(argv)).orderBook(request, callback(argv, displayTables));
};
