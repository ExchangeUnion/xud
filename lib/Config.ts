import fs from 'fs';
import os from 'os';
import path from 'path';
import toml from 'toml';
import utils from './utils/utils';
import { PoolConfig } from './p2p/Pool';
import { LndClientConfig } from './lndclient/LndClient';
import { RaidenClientConfig } from './raidenclient/RaidenClient';
import { DBConfig } from './db/DB';
import { OrderBookConfig } from './orderbook/OrderBook';

class Config {
  public p2p: PoolConfig;
  public xudir: string;
  public db: DBConfig;
  public testDb: DBConfig;
  public rpc: any;
  public lnd: LndClientConfig;
  public raiden: RaidenClientConfig;
  public orderbook: OrderBookConfig;

  constructor(private args?: object) {
    const platform = os.platform();
    let lndDatadir;
    switch (platform) {
      case 'win32': { // windows
        const homeDir = process.env.LOCALAPPDATA;
        this.xudir = `${homeDir}/Xud/`;
        lndDatadir = `${homeDir}/Lnd/`; // default lnd directory location
        break;
      }
      case 'darwin': { // mac
        const homeDir = process.env.HOME;
        this.xudir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/Library/Application Support/Lnd/`;
        break;
      }
      default: { // linux
        const homeDir = process.env.HOME;
        this.xudir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/.lnd/`;
        break;
      }
    }

    // default configuration
    this.p2p = {
      listen: true,
      port: 8885, // X = 88, U = 85 in ASCII
    };
    this.db = {
      host: 'localhost',
      port: 3306,
      username: 'xud',
      database: 'xud',
      dialect: 'mysql',
    };
    this.testDb = {
      ...this.db,
      database: 'xud_test',
    };
    this.rpc = {
      port: 8886,
    };
    this.lnd = {
      disable: false,
      datadir: lndDatadir,
      certpath: path.join(lndDatadir, 'tls.cert'),
      macaroonpath: path.join(lndDatadir, 'admin.macaroon'),
      host: 'localhost',
      port: 10009,
    };
    this.raiden = {
      disable: false,
      host: 'localhost',
      port: 5001,
    };
    this.orderbook = {
      internalmatching: true,
    };
  }

  public async load() {
    if (!fs.existsSync(this.xudir)) {
      fs.mkdirSync(this.xudir);
    } else if (fs.existsSync(`${this.xudir}xud.conf`)) {
      const configText: any = fs.readFileSync(`${this.xudir}xud.conf`);
      try {
        const props = toml.parse(configText);

        // merge parsed json properties from config file to this config object
        utils.deepMerge(this, props);
      } catch (e) {
        throw new Error(`Parsing error on line ${e.line}, column ${e.column
        }: ${e.message}`);
      }
    }

    if (this.args) {
      // override our config file with command line arguments
      utils.deepMerge(this, this.args);
      this.args = undefined;
    }
  }
}

export default Config;
