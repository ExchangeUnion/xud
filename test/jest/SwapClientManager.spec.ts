import Config from '../../lib/Config';
import { SwapClientType } from '../../lib/constants/enums';
import DB from '../../lib/db/DB';
import Logger from '../../lib/Logger';
import { OwnLimitOrder } from '../../lib/orderbook/types';
import { errorCodes } from '../../lib/swaps/errors';
import SwapClient from '../../lib/swaps/SwapClient';
import SwapClientManager from '../../lib/swaps/SwapClientManager';
import { UnitConverter } from '../../lib/utils/UnitConverter';
import { getUnitConverter } from '../utils';

jest.mock('../../lib/db/DB', () => {
  return jest.fn().mockImplementation(() => {
    return {
      models: {
        Currency: {
          findAll: () => {
            return [{ id: 'WETH', tokenAddress: '0x1234' }];
          },
        },
      },
    };
  });
});
jest.mock('../../lib/Config');
jest.mock('../../lib/Logger', () => {
  return jest.fn().mockImplementation(() => {
    return {
      createSubLogger: () => {},
      info: () => {},
    };
  });
});
jest.mock('../../lib/nodekey/NodeKey');
const mockLndPubKey = 1;
const lndInfoMock = jest.fn(() => Promise.resolve());
const onListenerMock = jest.fn();
const closeMock = jest.fn();
const mockLndOpenChannel = jest.fn();
jest.mock('../../lib/lndclient/LndClient', () => {
  return jest.fn().mockImplementation(() => {
    return {
      on: onListenerMock,
      init: () => Promise.resolve(),
      pubKey: mockLndPubKey,
      type: SwapClientType.Lnd,
      isDisabled: () => false,
      isOperational: () => true,
      isMisconfigured: () => false,
      getLndInfo: lndInfoMock,
      close: closeMock,
      openChannel: mockLndOpenChannel,
    };
  });
});
const tokenAddresses = new Map<string, string>();
jest.mock('../../lib/swaps/SwapClient');
const mockedSwapClient = <jest.Mock<SwapClient>>(<any>SwapClient);

const logger = new Logger({});
logger.error = jest.fn();
const loggers = {
  global: logger,
  db: logger,
  rpc: logger,
  p2p: logger,
  orderbook: logger,
  lnd: logger,
  connext: logger,
  swaps: logger,
  http: logger,
  service: logger,
};

describe('Swaps.SwapClientManager', () => {
  let config: Config;
  let db: DB;
  let swapClientManager: SwapClientManager;
  let unitConverter: UnitConverter;

  beforeEach(async () => {
    config = new Config();
    config.lnd = {
      BTC: {
        disable: false,
        certpath: 'tls.cert',
        host: 'localhost',
        port: 10009,
        nomacaroons: true,
        macaroonpath: '',
        cltvdelta: 40,
      },
      LTC: {
        disable: false,
        certpath: 'tls.cert',
        host: 'localhost',
        port: 10009,
        nomacaroons: true,
        macaroonpath: '',
        cltvdelta: 576,
      },
    };
    config.connext = {
      disable: false,
      host: 'localhost',
      port: 4321,
      webhookhost: 'localhost',
      webhookport: 4422,
    };
    config.strict = true;
    db = new DB(loggers.db, config.dbpath);
    unitConverter = getUnitConverter();
    tokenAddresses.set('WETH', '0x1234');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('it initializes lnd-ltc and lnd-btc', async () => {
    swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
    await swapClientManager.init();

    expect(swapClientManager['swapClients'].size).toEqual(2);
    expect(onListenerMock).toHaveBeenCalledTimes(6);
    expect(swapClientManager.get('BTC')).not.toBeUndefined();
    expect(swapClientManager.get('LTC')).not.toBeUndefined();
    expect(swapClientManager['swapClients'].size).toEqual(2);
    const lndClients = swapClientManager.getLndClientsMap();
    expect(lndClients.size).toEqual(2);
    expect(lndClients.get('BTC')!.pubKey).toEqual(1);
    expect(lndClients.get('LTC')!.pubKey).toEqual(1);
    await swapClientManager.getLndClientsInfo();
    expect(lndInfoMock).toHaveBeenCalledTimes(2);
  });

  test('it initializes lnd-ltc and lnd-btc', async () => {
    swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
    await swapClientManager.init();

    expect(swapClientManager['swapClients'].size).toEqual(2);
    expect(onListenerMock).toHaveBeenCalledTimes(6);
    expect(swapClientManager.get('BTC')).not.toBeUndefined();
    expect(swapClientManager.get('LTC')).not.toBeUndefined();
    swapClientManager.close();
    expect(closeMock).toHaveBeenCalledTimes(2);
  });

  test('it initializes lnd-btc', async () => {
    config.lnd.LTC!.disable = true;
    swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
    await swapClientManager.init();

    expect(swapClientManager['swapClients'].size).toEqual(1);
    expect(onListenerMock).toHaveBeenCalledTimes(3);
    expect(swapClientManager.get('BTC')).not.toBeUndefined();
    swapClientManager.close();
    expect(closeMock).toHaveBeenCalledTimes(1);
  });

  test('it initializes nothing', async () => {
    config.lnd.BTC!.disable = true;
    config.lnd.LTC!.disable = true;
    swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
    await swapClientManager.init();

    expect(swapClientManager['swapClients'].size).toEqual(0);
    expect(onListenerMock).toHaveBeenCalledTimes(0);
    expect(swapClientManager.get('BTC')).toBeUndefined();
    expect(swapClientManager.get('WETH')).toBeUndefined();
    swapClientManager.close();
    expect(closeMock).toHaveBeenCalledTimes(0);
  });

  test('closes lnd-btc and lnd-ltc', async () => {
    swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
    await swapClientManager.init();

    expect(swapClientManager['swapClients'].size).toEqual(2);
    swapClientManager.close();
    expect(closeMock).toHaveBeenCalledTimes(2);
  });

  describe('reserved amounts', () => {
    const currency = 'BTC';
    const amount = 10000;
    const setReservedInboundBtcAmount = jest.fn();

    beforeEach(async () => {
      swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
      await swapClientManager.init();
      swapClientManager.swapClients.get(currency)!.setReservedInboundAmount = setReservedInboundBtcAmount;
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('it adds outbound reserved amounts', () => {
      expect(swapClientManager.getOutboundReservedAmount(currency)).toBeUndefined();
      swapClientManager.addOutboundReservedAmount(currency, amount);
      expect(swapClientManager.getOutboundReservedAmount(currency)).toEqual(amount);
      swapClientManager.addOutboundReservedAmount(currency, amount);
      expect(swapClientManager.getOutboundReservedAmount(currency)).toEqual(amount * 2);
    });

    test('it subtracts outbound reserved amounts', () => {
      expect(swapClientManager.getOutboundReservedAmount(currency)).toBeUndefined();
      swapClientManager.addOutboundReservedAmount(currency, amount);
      expect(swapClientManager.getOutboundReservedAmount(currency)).toEqual(amount);
      swapClientManager.subtractOutboundReservedAmount(currency, amount);
      expect(swapClientManager.getOutboundReservedAmount(currency)).toEqual(0);
    });

    test('it adds inbound reserved amounts and sets amount on swap client', () => {
      expect(swapClientManager['inboundReservedAmounts'].get(currency)).toBeUndefined();
      swapClientManager.addInboundReservedAmount(currency, amount);
      expect(swapClientManager['inboundReservedAmounts'].get(currency)).toEqual(amount);
      expect(setReservedInboundBtcAmount).toHaveBeenLastCalledWith(amount, currency);
      swapClientManager.addInboundReservedAmount(currency, amount);
      expect(swapClientManager['inboundReservedAmounts'].get(currency)).toEqual(amount * 2);
      expect(setReservedInboundBtcAmount).toHaveBeenLastCalledWith(amount * 2, currency);
    });

    test('it subtracts inbound reserved amounts', () => {
      expect(swapClientManager['inboundReservedAmounts'].get(currency)).toBeUndefined();
      swapClientManager.addInboundReservedAmount(currency, amount);
      expect(swapClientManager['inboundReservedAmounts'].get(currency)).toEqual(amount);
      expect(setReservedInboundBtcAmount).toHaveBeenLastCalledWith(amount, currency);
      swapClientManager.subtractInboundReservedAmount(currency, amount);
      expect(swapClientManager['inboundReservedAmounts'].get(currency)).toEqual(0);
    });
  });

  describe('openChannel', () => {
    let remoteIdentifier: string;

    beforeEach(async () => {
      remoteIdentifier = '02afaef2634e5c7ca8d682b828a62bd040929b1e4b5030b21e2a0a891cf545b2e1';
      swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
      await swapClientManager.init();
    });

    test('it fails without swap client', async () => {
      expect.assertions(1);
      const currency = 'BTC';
      const amount = 16000000;
      swapClientManager.get = jest.fn().mockReturnValue(undefined);
      try {
        await swapClientManager.openChannel({
          remoteIdentifier,
          currency,
          amount,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('it fails without peerSwapClientPubKey', async () => {
      const currency = 'BTC';
      const amount = 16000000;
      try {
        await swapClientManager.openChannel({
          remoteIdentifier,
          currency,
          amount,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    test('it opens a channel using lnd', async () => {
      const currency = 'BTC';
      const amount = 16000000;
      const getClientSpy = jest.spyOn(swapClientManager, 'get');
      const lndListeningUris = ['123.456.789.321:9735', '192.168.63.155:9777'];
      await swapClientManager.openChannel({
        remoteIdentifier,
        currency,
        amount,
        uris: lndListeningUris,
      });
      expect(getClientSpy).toHaveBeenCalledWith(currency);
      expect(mockLndOpenChannel).toHaveBeenCalledTimes(1);
      expect(mockLndOpenChannel).toHaveBeenCalledWith(
        expect.objectContaining({
          remoteIdentifier,
          units: BigInt(amount),
          uris: lndListeningUris,
        }),
      );
    });
  });

  describe('tradingLimits & checkSwapCapacities', () => {
    beforeAll(() => {
      const btcClient = new mockedSwapClient();
      btcClient.isConnected = jest.fn().mockImplementation(() => true);
      btcClient.swapCapacities = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          maxOutboundChannelCapacity: 200000,
          maxInboundChannelCapacity: 150000,
          totalOutboundCapacity: 200000,
          totalInboundCapacity: 150000,
        });
      });
      swapClientManager.swapClients.set('BTC', btcClient);

      const ltcClient = new mockedSwapClient();
      ltcClient.isConnected = jest.fn().mockImplementation(() => true);
      ltcClient.swapCapacities = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          maxOutboundChannelCapacity: 700000,
          maxInboundChannelCapacity: 550000,
          totalOutboundCapacity: 700000,
          totalInboundCapacity: 550000,
        });
      });
      swapClientManager.swapClients.set('LTC', ltcClient);
    });

    test('returns trading limits for a currency', async () => {
      const btcTradingLimits = await swapClientManager.tradingLimits('BTC');

      expect(btcTradingLimits).toBeTruthy();
      expect(btcTradingLimits.maxSell).toEqual(200000);
      expect(btcTradingLimits.maxBuy).toEqual(150000);
    });

    test('throws if outbound swap capacity is insufficient for an order', async () => {
      const order: OwnLimitOrder = {
        price: 0.01,
        pairId: 'LTC/BTC',
        quantity: 800000,
        isBuy: false,
        localId: 'test',
      };

      await expect(swapClientManager.checkSwapCapacities(order)).rejects.toHaveProperty(
        'code',
        errorCodes.INSUFFICIENT_OUTBOUND_CAPACITY,
      );
    });

    test('throws if inbound swap capacity is insufficient for an order', async () => {
      const order: OwnLimitOrder = {
        price: 1,
        pairId: 'LTC/BTC',
        quantity: 300000,
        isBuy: false,
        localId: 'test',
      };

      await expect(swapClientManager.checkSwapCapacities(order)).rejects.toHaveProperty(
        'code',
        errorCodes.INSUFFICIENT_INBOUND_CAPACITY,
      );
    });

    test('throws if outbound swap capacity is insufficient for an order plus reserved amounts', async () => {
      const order: OwnLimitOrder = {
        price: 0.01,
        pairId: 'LTC/BTC',
        quantity: 600000,
        isBuy: false,
        localId: 'test',
      };

      swapClientManager['outboundReservedAmounts'].set('LTC', 200000);

      await expect(swapClientManager.checkSwapCapacities(order)).rejects.toHaveProperty(
        'code',
        errorCodes.INSUFFICIENT_OUTBOUND_CAPACITY,
      );
    });

    test('throws if inbound swap capacity is insufficient for an order plus reserved amounts', async () => {
      const order: OwnLimitOrder = {
        price: 1,
        pairId: 'LTC/BTC',
        quantity: 100000,
        isBuy: false,
        localId: 'test',
      };

      swapClientManager['inboundReservedAmounts'].set('BTC', 200000);

      await expect(swapClientManager.checkSwapCapacities(order)).rejects.toHaveProperty(
        'code',
        errorCodes.INSUFFICIENT_INBOUND_CAPACITY,
      );
    });

    test('returns if swap capacity is sufficient', async () => {
      const order: OwnLimitOrder = {
        price: 0.01,
        pairId: 'LTC/BTC',
        quantity: 600000,
        isBuy: false,
        localId: 'test',
      };

      swapClientManager['outboundReservedAmounts'].set('LTC', 50000);
      swapClientManager['inboundReservedAmounts'].set('BTC', 50000);

      await swapClientManager.checkSwapCapacities(order);
    });

    test('throws when swap client is not found', async () => {
      await expect(swapClientManager.tradingLimits('BBB')).rejects.toHaveProperty(
        'code',
        errorCodes.SWAP_CLIENT_NOT_FOUND,
      );

      const order: OwnLimitOrder = {
        price: 0.01,
        pairId: 'AAA/BBB',
        quantity: 600000,
        isBuy: false,
        localId: 'test',
      };

      await expect(swapClientManager.checkSwapCapacities(order)).rejects.toThrow(
        'cannot convert AAA amount of 600000 to units because decimal places per currency was not found in the database',
      );
    });
  });
});
