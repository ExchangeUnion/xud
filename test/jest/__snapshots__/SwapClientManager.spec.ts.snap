// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swaps.SwapClientManager openChannel it fails without swap client 1`] = `
Object {
  "code": "7.1",
  "message": "swapClient for currency BTC not found",
}
`;
